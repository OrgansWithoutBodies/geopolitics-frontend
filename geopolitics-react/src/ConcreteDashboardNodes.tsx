import {
  EventID,
  Network,
  Timeline,
  WorldMap,
} from "react-konva-components/src";
import { AdaptPlugin } from "./DashboardNodes";
import { IMapNode, INetworkNode, ITimelineNode } from "./DashboardNodes.types";

export function NetworkDashboardNode({
  inputs: {
    entities: { nodes, edges },
  },
  outputs: { selectedEntities },
  stageSize,
}: AdaptPlugin<INetworkNode<string>>): JSX.Element {
  return <Network nodes={nodes} stageSize={stageSize} edges={edges} />;
}
export function WorldMapDashboardNode({
  inputs: { entities, fills },
  outputs: { selectedEntities },
  stageSize,
}: AdaptPlugin<IMapNode<string>>): JSX.Element {
  return (
    <WorldMap
      // TODO select/onSelect
      container={{
        sizePx: stageSize,
        center: [],
      }}
      contents={{
        countries: entities,
        countryToRegion: undefined,
        countryToName: undefined,
        countryLines: undefined,
        countryHeartMap: undefined,
        highlights: fills,
      }}
    />
  );
}
export function TimelineDashboardNode({
  inputs: { entities },
  outputs: { selectedEntities },
  options: { timelineStart, timelineEnd },
  stageSize,
}: AdaptPlugin<ITimelineNode<number>>): JSX.Element {
  return (
    <Timeline
      stageSize={stageSize}
      onHoveredEvent={function (id: EventID): void {
        throw new Error("Function not implemented.");
      }}
      onSelectEvent={function (id: EventID): void {
        throw new Error("Function not implemented.");
      }}
      events={entities}
      latestEventEnd={timelineEnd}
      earliestEventStart={timelineStart}
      selectedEventId={null}
    />
  );
}
