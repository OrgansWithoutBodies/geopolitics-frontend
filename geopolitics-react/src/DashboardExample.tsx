// import { LinePlot, Network, Timeline, WorldMap } from "react-konva-components/src";
// import { ArrV2, HexString, TimeSpace } from "type-library";
// import { dataService } from "./data/data.service";
// import { CountryID } from "./data/data.store";
// import { MS_IN_YEAR, unoffsetDate } from "./timeTools";

// export function Dashboard1():JSX.Element{
//   return    <>
//   {filterYearsRenderReady !== undefined && (
//     <div>
//       <div>
//         <p>START YEAR</p>
//         <button
//           onClick={() =>
//             dataService.setFilterEndpoint(
//               "start",
//               filterYearsNullSafe.start + 1 * MS_IN_YEAR
//             )
//           }
//         >
//           +
//         </button>
//         <input
//           value={filterYearsRenderReady.start}
//           onChange={(val) => {
//             dataService.setFilterEndpoint(
//               "start",
//               unoffsetDate(Number.parseInt(val.target.value) as TimeSpace)
//             );
//           }}
//         />
//         <button
//           onClick={() =>
//             dataService.setFilterEndpoint(
//               "start",
//               filterYearsNullSafe.start - 1 * MS_IN_YEAR
//             )
//           }
//         >
//           -
//         </button>
//       </div>
//       <div>
//         <p>END YEAR</p>
//         <button
//           onClick={() =>
//             dataService.setFilterEndpoint(
//               "end",
//               filterYearsNullSafe.end + 1 * MS_IN_YEAR
//             )
//           }
//         >
//           +
//         </button>
//         <input
//           value={filterYearsRenderReady.end}
//           onChange={(val) => {
//             dataService.setFilterEndpoint(
//               "end",
//               unoffsetDate(Number.parseInt(val.target.value) as TimeSpace)
//             );
//           }}
//         />
//         <button
//           onClick={() =>
//             dataService.setFilterEndpoint(
//               "end",
//               filterYearsNullSafe.end - 1 * MS_IN_YEAR
//             )
//           }
//         >
//           -
//         </button>
//       </div>
//       </div>
//   )}
//   <LinePlot
//     stageSize={{ x: COLUMN_1_WIDTH, y: 200 }}
//     data={[
//       {
//         points: [...Array(100).keys()].map(
//           (ii) => [ii / 2, 10 * Math.sin(ii / 2)] as ArrV2
//         ),
//         color: highlightedPlot === 0 ? "#FFFF00" : ("#FF0000" as HexString),
//       },
//       {
//         points: [...Array(100).keys()].map(
//           (ii) => [ii / 2, -10 * Math.sin(ii / 2)] as ArrV2
//         ),
//         color: highlightedPlot === 1 ? "#FFFF00" : ("#0000FF" as HexString),
//       },
//     ]}
//     centerPoint={{ x: -COLUMN_1_WIDTH / 2, y: 0 }}
//     // centerPoint={{ x: 0, y: 0 }}
//     plotRange={{
//       min: {
//         x: 0,
//         y: -10,
//       },
//       max: {
//         x: 100,
//         y: 10,
//       },
//     }}
//     onSelect={setHighlightedPlot}
//   />
//   {/* <SliderBar
//     handlePercs={{ low: 0, high: 1 }}
//     stageSize={{
//       x: COLUMN_1_WIDTH,
//       y: TIMELINE_HEIGHT / 2,
//     }}
//   /> */}
//   <div style={{ display: "flex", flexDirection: "row" }}>
//     <div style={{ display: "flex", flexDirection: "column" }}>
//       {/* <QuotedText /> */}
//       {countryToName && countries && (
//         <div>
//           <WorldMap
//             container={{
//               sizePx: { x: COLUMN_1_WIDTH, y: MAP_HEIGHT },
//               center: [0, 0],
//             }}
//             contents={{
//               countries,
//               countryToName,
//               countryHeartMap,
//               countryNodeColors: nodeColorLookup,
//               countryLines: bilateralRelations,
//               onClick: (id) =>
//                 dataService.setSelectedCountry(id as CountryID),
//               highlights: [
//                 {
//                   highlightColor: "#FFFF00",
//                   highlightedCountries:
//                     selectedCountry !== null ? [selectedCountry] : [],
//                 },
//               ],
//             }}
//           />
//         </div>
//       )}
//       {countryStarts && (
//         <Timeline
//           // can only switch ID type bc rn we're using the same code, terrible pattern TODO
//           onEventClick={(eventId) => {
//             dataService.setSelectedCountry(eventId as any as CountryID);
//           }}
//           stageSize={{
//             x: COLUMN_1_WIDTH,
//             y: TIMELINE_HEIGHT,
//           }}
//           events={countryStarts}
//           // we pass in "MS since Jan 1 1970"
//         />
//       )}
//     </div>
//     {true && (
//       <Network
//         nodes={nodes}
//         edges={edges}
//         stageSize={{
//           x: COLUMN_2_WIDTH,
//           y: TIMELINE_HEIGHT + MAP_HEIGHT,
//         }}
//         NodeTemplate={NodeTemplate}
//       />
//     )
//     }
//     </>
// }
