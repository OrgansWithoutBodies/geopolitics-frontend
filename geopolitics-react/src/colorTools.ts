/**
 * TODO - refactor this as a (jQuery?) plugin!
 **/

import { HexString } from "type-library";
import { HexTripleNumber } from "./types";

// Converts a #ffffff hex string into an [r,g,b] array
function h2r(hex: HexString): HexTripleNumber | null {
  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  return result
    ? ([
        parseInt(result[1], 16),
        parseInt(result[2], 16),
        parseInt(result[3], 16),
      ] as HexTripleNumber)
    : null;
}

// Inverse of the above
function r2h(rgb: HexTripleNumber): HexString {
  return `#${((1 << 24) + (rgb[0] << 16) + (rgb[1] << 8) + rgb[2])
    .toString(16)
    .slice(1)}`;
}

// Interpolates two [r,g,b] colors and returns an [r,g,b] of the result
// Taken from the awesome ROT.js roguelike dev library at
// https://github.com/ondras/rot.js
function _interpolateColor(
  color1: HexTripleNumber,
  color2: HexTripleNumber,
  factor = 0.5
): HexTripleNumber {
  const result = color1.map((_, ii) =>
    Math.round(color1[ii] + factor * (color2[ii] - color1[ii]))
  ) as HexTripleNumber;
  return result;
}

// const rgb2hsl = function (color: HexTripleNumber): HexTripleNumber {
//   const r = color[0] / 255;
//   const g = color[1] / 255;
//   const b = color[2] / 255;

//   const max = Math.max(r, g, b),
//     min = Math.min(r, g, b);
//   let h: number = (max + min) / 2;
//   let s: number = h;
//   const l: number = h;
//   if (max == min) {
//     h = 0; // achromatic
//     s = 0;
//   } else {
//     const d = max - min;
//     s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
//     switch (max) {
//       case r:
//         h = (g - b) / d + (g < b ? 6 : 0);
//         break;
//       case g:
//         h = (b - r) / d + 2;
//         break;
//       case b:
//         h = (r - g) / d + 4;
//         break;
//     }
//     h /= 6;
//   }

//   return [h, s, l] as HexTripleNumber;
// };

// function hue2rgb(p: number, q: number, t: number) {
//   if (t < 0) t += 1;
//   if (t > 1) t -= 1;
//   if (t < 1 / 6) return p + (q - p) * 6 * t;
//   if (t < 1 / 2) return q;
//   if (t < 2 / 3) return p + (q - p) * (2 / 3 - t) * 6;
//   return p;
// }

// const hsl2rgb = function (color: HexTripleNumber) {
//   let l = color[2];

//   if (color[1] == 0) {
//     l = Math.round(l * 255) as HexTripleNumber[number];
//     return [l, l, l];
//   } else {
//     const s = color[1];
//     const q = l < 0.5 ? l * (1 + s) : l + s - l * s;
//     const p = 2 * l - q;
//     const r = hue2rgb(p, q, color[0] + 1 / 3);
//     const g = hue2rgb(p, q, color[0]);
//     const b = hue2rgb(p, q, color[0] - 1 / 3);
//     return [Math.round(r * 255), Math.round(g * 255), Math.round(b * 255)];
//   }
// };

// const _interpolateHSL = function (
//   color1: HexTripleNumber,
//   color2: HexTripleNumber,
//   factor: number
// ) {
//   if (arguments.length < 3) {
//     factor = 0.5;
//   }
//   const hsl1 = rgb2hsl(color1);
//   const hsl2 = rgb2hsl(color2);
//   for (let i = 0; i < 3; i++) {
//     hsl1[i] += (factor * (hsl2[i] - hsl1[i])) as HexTripleNumber[number];
//   }
//   return hsl2rgb(hsl1);
// };

export function interpolateHexStrings(
  hexStringA: HexString,
  hexStringB: HexString
): HexString {
  const tripleA = h2r(hexStringA);
  const tripleB = h2r(hexStringB);
  if (tripleA === null || tripleB === null) {
    return "#FFFFFF";
  }
  console.log("TEST123", tripleA, tripleB);
  const intermediatePoint = _interpolateColor(tripleA, tripleB);
  return r2h(intermediatePoint);
}
