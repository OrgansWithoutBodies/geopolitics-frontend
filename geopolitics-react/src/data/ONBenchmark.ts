import * as asciichart from "asciichart";

import PolynomialRegression from "ml-regression-polynomial";

import Bench, { TaskResult } from "tinybench";

/**
 * "O(N) Benchmark"
 * @param nFunc
 * @param minN
 * @param maxN
 * @param stepN
 * @returns
 */
export async function ONBenchmark<TSetup>({
  nFunc,
  setupFunc,
  minN,
  maxN,
  numSteps = 100,
  iterations = 30,
}: {
  nFunc: (setupParams: TSetup) => void;
  setupFunc: (N: number) => TSetup;

  minN: number;
  maxN: number;
  numSteps?: number;
  iterations?: number;
}) {
  const stepN = Math.floor((maxN - minN) / numSteps);
  let N = minN;
  const results: [number, TaskResult][] = [];
  // const minMax: {min:number,max:number}[] = [];
  while (N < maxN) {
    // set up params before running bench
    const setupParams = setupFunc(N);

    const bench = new Bench({ time: 100, iterations });
    bench.add("bench", async () => {
      nFunc(setupParams);
    });
    await bench.run();
    results.push([N, bench.results[0]!]);
    // minMax.push([N, bench.results[0]!.mean]);
    N = N + stepN;
  }
  return results;
}

export async function performBenchTest<TSetupType>({
  nFunc,
  setupFunc,
  minN = 1,
  maxN = 100,
  numSteps = 100,
  degree = 2,
  iterations = 30,
}: TSetupType extends (setupParams: infer TSetup) => any
  ? {
      nFunc: TSetupType;
      setupFunc: (N: number) => TSetup;
      minN?: number;
      maxN?: number;
      numSteps?: number;
      degree?: number;
      iterations?: number;
    }
  : never): Promise<JSONType<"polynomialRegression">> {
  const results = await ONBenchmark({
    nFunc,
    setupFunc,
    minN,
    maxN,
    numSteps,
    iterations,
  });
  console.table(
    Object.fromEntries(results.map((val) => [val[0], val[1].mean]))
  );
  const means = results.map((val) => val[1].mean);
  // const regression = linearRegression(
  //   results.map((val) => [val[0], val[1].mean])
  // );
  const regression = new PolynomialRegression(
    results.map((val) => val[0]),
    results.map((val) => val[1].mean),
    degree
  ).toJSON();
  const maxMean = Math.max(...means);
  const minMean = Math.min(...means);
  console.log(regression);
  console.log(
    asciichart.plot(
      [
        means.map((_, ii) => {
          // return ii;
          return ((maxMean - minMean) / means.length) * ii;
        }),
        means,
      ],
      {
        colors: [asciichart.red, asciichart.blue],
        height: 20,
      }
    )
  );

  console.log(
    asciichart.plot([results.map((val) => val[1].sd)], {
      colors: [asciichart.green],
      height: 10,
    })
  );
  console.log("Blue Function, Red N, Green SD");
  return regression;
}
