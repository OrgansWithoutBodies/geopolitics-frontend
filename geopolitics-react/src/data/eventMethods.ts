import {
  periodIsSegmentGuard,
  type EventID,
  type HistoricalEvent,
  type LineSegment,
  type PeriodOrSingleton,
  type SpaceConvertingFunction,
} from "react-konva-components/src/types";
import type { TimeSpace, TimelineSpace } from "type-library";
import type { CountryID, CountryType } from "./data.store";
import { numericalQCode } from "./numericalQCode";

const timeObjAAfterTimeObjB = (
  timeA: PeriodOrSingleton<TimeSpace>,
  timeB: PeriodOrSingleton<TimeSpace>,

  key: keyof LineSegment
): boolean => {
  const aIsSegment = periodIsSegmentGuard(timeA);
  const bIsSegment = periodIsSegmentGuard(timeB);

  if (aIsSegment && bIsSegment) {
    return timeA[key] >= timeB[key];
  }
  if (aIsSegment && !bIsSegment) {
    return timeA[key] >= timeB;
  }
  if (!aIsSegment && bIsSegment) {
    return timeA >= timeB[key];
  }
  if (!aIsSegment && !bIsSegment) {
    return timeA >= timeB;
  }
  return false;
};

export function applyFiltersToEvents(
  initialDateFilter: TimeSpace | null,
  events: HistoricalEvent[],
  finalDateFilter: TimeSpace | null
) {
  if (!(initialDateFilter !== null || finalDateFilter !== null)) {
    return events;
  }

  const filteredEvents = events.filter((event) => {
    const afterInitial =
      initialDateFilter === null
        ? true
        : timeObjAAfterTimeObjB(event.eventTime, initialDateFilter, "start");
    const beforeFinal =
      finalDateFilter === null
        ? true
        : timeObjAAfterTimeObjB(finalDateFilter, event.eventTime, "end");
    return afterInitial && beforeFinal;
  });

  return filteredEvents;
}

export const sortEvents = (
  events: HistoricalEvent[],
  sortKey: keyof LineSegment
) => {
  // console.log("TODO", sortKey);
  return [...events].sort((eventA, eventB) => {
    return timeObjAAfterTimeObjB(eventA.eventTime, eventB.eventTime, sortKey)
      ? 1
      : -1;
  });
};

// TODO is timeline len supposed to be number of years?
export const buildEventPositioner = (
  earliestDate: TimeSpace,
  latestDate: TimeSpace
) => {
  const timelinePositionForEvent: SpaceConvertingFunction<
    TimeSpace,
    TimelineSpace
  > = (val) => {
    const timelineLen = latestDate - earliestDate;
    return ((val - earliestDate) / timelineLen) as TimelineSpace;
  };

  const eventPositioner = ({ eventTime }: HistoricalEvent): TimelineSpace => {
    const isSegment = periodIsSegmentGuard(eventTime);
    if (isSegment) {
      // for time period events just position tooltip in the middle
      return ((timelinePositionForEvent(eventTime.end) +
        timelinePositionForEvent(eventTime.start)) /
        2) as TimelineSpace;
    }

    return timelinePositionForEvent(eventTime);
  };
  return eventPositioner;
};
export function positionNode(
  country: CountryType,
  start: TimeSpace,
  end: TimeSpace,
  qCodes: Record<`Q${CountryID}`, string>,
  selectedCountry: CountryID | null,
  nodeColorLookup: Record<`${CountryID}`, `#${string}`>
) {
  const startDate = country.stateStart.value;
  const eventTime = new Date(startDate).getTime() as TimeSpace;
  const positioner = buildEventPositioner(start, end);
  const genericProps = {
    id: numericalQCode(country) as EventID,
    eventName: `Beginning of ${qCodes[country.item.value]}`,
    eventInfo: "",
    eventTime,
  };
  return {
    ...genericProps,
    renderedProps: {
      // TODO no as any, same as overloading country & event id as same
      color:
        (selectedCountry as any) === (genericProps.id as any)
          ? "#FFFF00"
          : nodeColorLookup[`${genericProps.id as any}`] || "#AAAAAA",
      position: positioner(genericProps),
    },
  };
}
