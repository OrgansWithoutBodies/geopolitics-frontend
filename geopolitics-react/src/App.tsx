import { ImperativePanelHandle } from "react-resizable-panels";
import "./App.css";

import { useEffect, useRef, useState } from "react";
import {
  CountryCode,
  CountryInfoKey,
  CountryNameLookup,
  CountryRegionLookup,
  RegionColorMap,
} from "../dataPrep/src/mapTypes";
import { HighlightSpecification } from "./WorldMap";
import { interpolateHexStrings } from "./colorTools";
import { countryInfo } from "./countryData";
import { ObjV2 } from "./types";
const allCountryCodes = countryInfo.map((val) => val["alpha-3"]);

const intersectSets = (a: Set<any>, b: Set<any>) =>
  new Set([...a].filter((i) => b.has(i)));
const notSet = (a: Set<any>, U: Set<any>) =>
  new Set([...U].filter((i) => !a.has(i)));
// subtract a - b
const subtractSets = (a: Set<any>, b: Set<any>, U: Set<any>) =>
  intersectSets(a, notSet(b, U));
``;

type HighlightRecord<TKey extends string> = Record<
  TKey,
  HighlightSpecification<CountryCode>
>;

function IntersectPartitions<TKeyA extends string, TKeyB extends string>(
  partitionA: HighlightRecord<TKeyA>,
  partitionB: HighlightRecord<TKeyB>,
  nameA: Readonly<string> = "A",
  nameB: Readonly<string> = "B"
): HighlightRecord<
  `${TKeyA}-${TKeyB}` | `${typeof nameA}:${TKeyA}` | `${typeof nameB}:${TKeyB}`
> {
  const allPartitionA = Object.values(partitionA)
    .map(
      (val) => (val as HighlightSpecification<CountryCode>).highlightedCountries
    )
    .flat() as CountryCode[];
  const allPartitionB = Object.values(partitionB)
    .map(
      (val) => (val as HighlightSpecification<CountryCode>).highlightedCountries
    )
    .flat() as CountryCode[];

  const partitionsFromIntersections = Object.keys(partitionA)
    .map((valA) => {
      return Object.keys(partitionB).map((valB) => {
        return [
          `${valA}-${valB}`,
          {
            highlightedCountries: getIntersectionsBetween(
              valA,
              valB,
              partitionA,
              partitionB
            ),
            highlightColor: interpolateHexStrings(
              partitionA[valA as TKeyA].highlightColor,
              partitionB[valB as TKeyB].highlightColor
            ),
          },
        ];
      });
    })
    .flat();

  const entriesInAButNotB = Object.keys(partitionA).map((valA) => [
    `${nameA}:${valA}`,
    {
      highlightedCountries: getSubtraction(
        partitionA[valA as TKeyA].highlightedCountries as CountryCode[],
        allPartitionB
      ),
      highlightColor: partitionA[valA as TKeyA].highlightColor,
    },
  ]);
  const entriesInBButNotA = Object.keys(partitionB).map((valB) => [
    `${nameB}:${valB}`,
    {
      highlightedCountries: getSubtraction(
        partitionB[valB as TKeyB].highlightedCountries as CountryCode[],
        allPartitionA
      ),
      highlightColor: partitionB[valB as TKeyB].highlightColor,
    },
  ]);

  return Object.fromEntries([
    ...partitionsFromIntersections,
    ...entriesInAButNotB,
    ...entriesInBButNotA,
  ]);
}
function App() {
  // const ref = useRef<ImperativePanelGroupHandle>(null);

  const resetLayout = () => {
    const panelGroup = ref.current;
    if (panelGroup) {
      // Reset each Panel to 50% of the group's width
      panelGroup.setLayout([50, 50]);
    }
  };

  const mapVals = {
    country: {
      sourceKey: "item",
      pCode: "P17",
      valueKey: "?country",
      joinChar: ".",
      optional: false,
    },
    coords: {
      sourceKey: "country",
      pCode: "P625",
      valueKey: "?coords",
      joinChar: ".",
      optional: false,
    },
    ideology: {
      sourceKey: "item",
      pCode: "P1142",
      valueKey: "?ideology",
      joinChar: ".",
      optional: true,
    },
    start: {
      sourceKey: "item",
      pCode: "P571",
      valueKey: "?start",
      joinChar: ".",
      optional: true,
    },
    end: {
      sourceKey: "item",
      pCode: "P576",
      valueKey: "?end",
      joinChar: ".",
      optional: true,
    },
  };
  type MembershipStatus = "current" | "applied" | "interest";
  const timelinePanelRef = useRef<ImperativePanelHandle>(null);
  const [paneSize, setPaneSize] = useState<number | null>(null);
  useEffect(() => {
    if (timelinePanelRef.current) {
      setPaneSize(timelinePanelRef.current.getSize());
    }
  }, [timelinePanelRef.current?.getSize]);
  const canvasSize: ObjV2 = { x: 1000, y: 300 };
  const regionColorMap: RegionColorMap = {
    Africa: "white",
    Americas: "white",
    Antarctica: "white",
    Asia: "white",
    Europe: "white",
    Oceania: "white",
  };
  // const regionColorMap: RegionColorMap = {
  //   Africa: "green",
  //   Americas: "yellow",
  //   Antarctica: "white",
  //   Asia: "blue",
  //   Europe: "cyan",
  //   Oceania: "red",
  // };
  const [year, setYear] = useState(2000);
  const mapCountryData = (key: CountryInfoKey, value: CountryInfoKey) =>
    Object.fromEntries(
      countryInfo.map((country) => [country[key], country[value]])
    );
  const countryToRegion = mapCountryData(
    "alpha-3",
    "region"
  ) as CountryRegionLookup<CountryCode>;
  const countryToName = mapCountryData(
    "alpha-3",
    "name"
  ) as CountryNameLookup<CountryCode>;
  return (
    <>
      <div>
        <div onClick={() => setYear(year + 1)}>+</div>
        <div onClick={() => setYear(year - 1)}>-</div>
      </div>
      {/* <QuotedText /> */}
      {/* <WorldMap<CountryCode>
        container={{
          sizePx: { x: 1024, y: 780 },
          center: [0, 0],
        }}
        contents={{
          countries: Object.entries(CountryFiles).map(([key, geometry]) => ({
            key,
            geometry,
          })),


          ],
          countryToRegion,
          countryToName,
          regionColorMap,
          // bilateralRelations,
          countryHeartMap: Capitals,
        }}
      /> */}
    </>
    // <div style={{ width: canvasSize.x, height: canvasSize.y }}>
    //   <div className={styles.Container}>
    //     <PanelGroup direction="vertical">
    //       <Panel
    //         defaultSize={100}
    //         className={styles.Panel}
    //         ref={timelinePanelRef}
    //       >
    //         <div className={styles.PanelContent}>
    //           <Timeline stageSize={{ x: canvasSize.x, y: paneSize || 0 }} />
    //         </div>
    //         {/* TEST */}
    //       </Panel>
    //       <ResizeHandle />
    //       <Panel className={styles.Panel}>
    //         <div className={styles.PanelContent}>
    //           <Network stageSize={{ x: canvasSize.x, y: paneSize || 0 }} />
    //         </div>
    //         {/* <Timeline /> */}
    //         {/* <AddNewEvent /> */}
    //       </Panel>
    //     </PanelGroup>
    //   </div>
    // </div>
  );
}
function getIntersectionsBetween(
  keyA: keyof typeof partitionA,
  keyB: keyof typeof partitionB,
  partitionA,
  partitionB
): CountryCode[] {
  return [
    ...intersectSets(
      new Set(partitionA[keyA].highlightedCountries),
      new Set(partitionB[keyB].highlightedCountries)
    ),
  ];
}
function getSubtraction(
  aList: CountryCode[],
  subList: CountryCode[]
): CountryCode[] {
  return [
    ...subtractSets(new Set(aList), new Set(subList), new Set(allCountryCodes)),
  ];
}

export default App;
const PanelStyles: Record<string, React.CSSProperties> = {
  container: {
    height: "100%",
    width: "100%",
    display: "flex",
    flexDirection: "row",
    alignItems: " center",
    justifyContent: " center",
    overflow: "hidden",
    borderRadius: " 0.5rem",
  },
};
