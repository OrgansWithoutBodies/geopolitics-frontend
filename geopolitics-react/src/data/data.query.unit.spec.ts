import { EventID, HistoricalEvent } from "react-konva-components/src";
import { TimeSpace } from "type-library";
import { performBenchTest } from "./ONBenchmark";
import { applyFiltersToEvents } from "./eventMethods";

// https://github.com/jiayihu/pretty-algorithms
describe("DataQuery Static Methods", () => {
  // it("applyFiltersToEvents Snapshot", () => {
  //   const N = 11;
  //   const filteredEvents = filter(N);
  //   // expect(events.length).toEqual(N);
  //   expect(filteredEvents.length).toEqual(9);
  //   // expect(filteredEvents).toMatchSnapshot();
  // });
  it("applyFiltersToEvents Benchmark", async () => {
    const {
      coefficients: [offset, slope, square],
    } = await performBenchTest<(val: HistoricalEvent[]) => void>({
      nFunc: (events) =>
        applyFiltersToEvents(10 as TimeSpace, events, 90 as TimeSpace),
      setupFunc: (N) =>
        generateEvents(N, () => Math.round(Math.random() * 100) as TimeSpace),
      minN: 1,
      maxN: 1000,
      numSteps: 40,
    });
    // 0.002
    expect(offset).toBeCloseTo(0, 1);
    expect(slope).toBeCloseTo(2e-3, 2);
    expect(square).toBeCloseTo(0, 6);
  });
  // it("applyFiltersToEvents Benchmark", async () => {
  //   const { m: slope } = await performBenchTest(
  //     (N) =>
  //       [...new Array(N).keys()].map((val) => [...new Array(N + val).keys()]),
  //     1,
  //     1000,
  //     20
  //   );
  //   expect(slope).toBeCloseTo(0.002, 3);
  // });
});

// function filter(
//   N: number,
//   eventTime: (N: number, ii: number) => TimeSpace = (N, ii) =>
//     ((ii * 100) / (N - 1)) as TimeSpace
// ) {
//   const events: HistoricalEvent[] = generateEvents(N, eventTime);
//   const filteredEvents = applyFiltersToEvents(
//     10 as TimeSpace,
//     events,
//     90 as TimeSpace
//   );
//   return filteredEvents;
// }

function generateEvents(
  N: number,
  getEventTime: (N: number, ii: number) => TimeSpace
): HistoricalEvent[] {
  return [...Array(N).keys()].map((ii) => ({
    id: ii as EventID,
    eventTime: getEventTime(N, ii),
    eventName: `Test${ii}-name`,
    eventInfo: `Test${ii}-info`,
  }));
}
