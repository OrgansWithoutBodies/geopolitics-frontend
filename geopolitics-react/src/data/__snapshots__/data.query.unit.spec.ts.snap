// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataQuery Static Methods applyFiltersToEvents Snapshot 1`] = `
[
  {
    "eventInfo": "Test1-info",
    "eventName": "Test1-name",
    "eventTime": 10,
    "id": 1,
  },
  {
    "eventInfo": "Test2-info",
    "eventName": "Test2-name",
    "eventTime": 20,
    "id": 2,
  },
  {
    "eventInfo": "Test3-info",
    "eventName": "Test3-name",
    "eventTime": 30,
    "id": 3,
  },
  {
    "eventInfo": "Test4-info",
    "eventName": "Test4-name",
    "eventTime": 40,
    "id": 4,
  },
  {
    "eventInfo": "Test5-info",
    "eventName": "Test5-name",
    "eventTime": 50,
    "id": 5,
  },
  {
    "eventInfo": "Test6-info",
    "eventName": "Test6-name",
    "eventTime": 60,
    "id": 6,
  },
  {
    "eventInfo": "Test7-info",
    "eventName": "Test7-name",
    "eventTime": 70,
    "id": 7,
  },
  {
    "eventInfo": "Test8-info",
    "eventName": "Test8-name",
    "eventTime": 80,
    "id": 8,
  },
  {
    "eventInfo": "Test9-info",
    "eventName": "Test9-name",
    "eventTime": 90,
    "id": 9,
  },
]
`;
